forest/shared/src/cli/mod.rs

```diff
@@ -176,60 +174,150 @@ impl CliOpts {
     if let Some(dir) = ProjectDirs::from("com", "ChainSafe", "Forest") {
-        let mut config_dir = dir.config_dir().to_path_buf();
-        config_dir.push("config.toml");
-        if config_dir.exists() {
-            info!("Found a config file at {}", config_dir.display());
-            return read_config_or_none(config_dir);
+        let path = dir.config_dir().join("config.toml");
+        if path.exists() {
+            return Some(ConfigPath::Project(path));
         }
     }
-
-    warn!("No configurations found, using defaults.");
-
     None
 }
 
-fn read_config_or_none(path: PathBuf) -> Option<Config> {
-    let toml = match read_file_to_string(&path) {
-        Ok(t) => t,
-        Err(e) => {
-            warn!("An error occured while reading configuration file at {}. Resorting to default configuration. Error was {}", path.display(), e);
-            return None;
+fn find_unknown_keys<'a>(
+    table: Option<&'a str>,
+    x: &'a toml::Value,
+    y: &'a toml::Value,
+    result: &mut Vec<(Option<&'a str>, &'a str)>,
+) {
+    if let toml::Value::Table(x_map) = x {
+        if let toml::Value::Table(y_map) = y {
+            let x_set: HashSet<_> = x_map.keys().collect();
+            let y_set: HashSet<_> = y_map.keys().collect();
+            for k in x_set.difference(&y_set) {
+                result.push((table, k));
+            }
+            for (x_key, x_value) in x_map.iter() {
+                if let Some(y_value) = y_map.get(x_key) {
+                    find_unknown_keys(Some(x_key), x_value, y_value, result);
+                }
+            }
+        }
+    }
+    if let toml::Value::Array(x_vec) = x {
+        if let toml::Value::Array(y_vec) = y {
+            for (x_value, y_value) in x_vec.iter().zip(y_vec.iter()) {
+                find_unknown_keys(table, x_value, y_value, result);
+            }
         }
-    };
-
-    match read_toml(&toml) {
-        Ok(cfg) => Some(cfg),
-        Err(e) => {
-            warn!(
-                "Error reading configuration file, opting to defaults. Error was {} ",
-                e
-            );
-            None
+    }
+}
+
+pub fn warn_for_unknown_keys(path: &Path, config: &Config) {
+    // `config` has been loaded successfully from toml file in `path` so we can always serialize
+    // it back to a valid TOML value or get the TOML value from `path`
+    let file = read_file_to_string(path).unwrap();
+    let value = file.parse::<toml::Value>().unwrap();
+
+    let config_file = toml::to_string(config).unwrap();
+    let config_value = config_file.parse::<toml::Value>().unwrap();
+
+    let mut result = vec![];
+    find_unknown_keys(None, &value, &config_value, &mut result);
+    for (table, k) in result.iter() {
+        let path = path.display();
+        if let Some(table) = table {
+            warn!("{path}: Unknown key `{k}` in [{table}], ignoring");
+        } else {
+            warn!("{path}: Unknown key `{k}` in top-level table, ignoring");
         }
```

Comment:
The message could be perhaps more precise, currently if we e.g. put a random new key in log config, i.e.
```
[[log.filters]]
module = "libp2p_bitswap"
level = "WARN"
cthulhu = "yes"
```
we only get `Unknown key `cthulhu` in [filters], ignoring`. This should rather be `[log.filters]`.

Reply:
Good idea!